postgres/postgres.go

```diff
@@ -0,0 +1,207 @@
+	row := ds.addS3Url.QueryRow(message.AliasName, message.URL, instanceID)
+	var returnedInstanceID sql.NullString
+	// Check that a instanceID is returned if not, no rows where update so return a job not found error
+	error := row.Scan(&returnedInstanceID)
+	return convertError(error)
+}
+
+// UpdateJobState - Update the state of a job.
+func (ds Datastore) UpdateJobState(jobID string, state *models.JobState) error {
+	row := ds.updateState.QueryRow(state.State, jobID)
+	var jobIDReturned sql.NullString
+	// Check that a instanceId is returned if not, no rows where update so return a job not found error
+	dataStoreError := row.Scan(&jobIDReturned)
+	return dataStoreError
+}
+
+// AddInstance - Add an instance and relate it to a job.
+func (ds Datastore) AddInstance(jobID string) (string, error) {
+	job := models.JobInstanceState{State: "Created", LastUpdated: time.Now().UTC().String(), Events: []models.Event{}}
+	bytes, error := json.Marshal(job)
+	if error != nil {
+		return "", error
+	}
+	row := ds.addInstance.QueryRow(jobID, bytes)
+	var instanceID sql.NullString
+	rowError := row.Scan(&instanceID)
+	if rowError != nil {
+		return "", rowError
+	}
+	return instanceID.String, nil
+}
+
+// GetInstance - Get an instance from postgres.
+func (ds Datastore) GetInstance(instanceID string) (models.JobInstanceState, error) {
+	row := ds.findInstance.QueryRow(instanceID)
+	var job sql.NullString
+	rowError := row.Scan(&job)
+	if rowError != nil {
+		return models.JobInstanceState{}, convertError(rowError)
+	}
+	var importJob models.JobInstanceState
+	error := json.Unmarshal([]byte(job.String), &importJob)
+	if error != nil {
+		return models.JobInstanceState{}, error
+	}
+	importJob.InstanceID = instanceID
+	return importJob, nil
+}
+
+// AddEvent - Add an event into an instance.
+func (ds Datastore) AddEvent(instanceID string, event *models.Event) error {
+	row := ds.addEvent.QueryRow(event.Type, event.Time, event.Message, event.MessageOffset, instanceID)
+	var returnedInstanceID sql.NullString
+	// Check that a instanceID is returned if not, no rows where update so return a job not found error
+	error := row.Scan(&returnedInstanceID)
+	return convertError(error)
+}
+
+// AddDimension - Add a dimension to cache in postgres
+func (ds Datastore) AddDimension(instanceID string, dimension *models.Dimension) error {
+	// Check that an instance exists else return an error
+	_, err := ds.GetInstance(instanceID)
+	if err != nil {
+		return err
+	}
+	res, queryError := ds.addDimension.Query(instanceID, dimension.NodeName, dimension.Value)
+	if queryError != nil {
+		return queryError
+	}
+	return res.Close()
+}
+
+// GetDimension - Get all dimensions related to an instanceID
+func (ds Datastore) GetDimension(instanceID string) ([]models.Dimension, error) {
+	_, err := ds.GetInstance(instanceID)
+	if err != nil {
+		return []models.Dimension{}, err
+	}
+	rows, err := ds.getDimensions.Query(instanceID)
```

Comment:
If a scan is called on the results the connection is automatically closed. In this case closed must be called as no scan happens